habitat/core/dataset.py

```diff
@@ -355,9 +367,65 @@ def _shuffle_iterator(self) -> None:
                 list(g)
                 for k, g in groupby(self._iterator, key=lambda x: x.scene_id)
             ]
-            random.shuffle(grouped_episodes)
+            grouped_episodes += []
+
+            for i in range(len(grouped_episodes)):
+                random.shuffle(grouped_episodes[i])
+
+            if len(grouped_episodes) > 1:
+                # Ensure we swap episodes by moving the current group to the end
+                random.shuffle(grouped_episodes[1:])
+                grouped_episodes = grouped_episodes[1:] + grouped_episodes[0:1]
+
             self._iterator = iter(sum(grouped_episodes, []))
         else:
             episodes = list(self._iterator)
             random.shuffle(episodes)
             self._iterator = iter(episodes)
+
+    def step_taken(self):
+        self._step_count += 1
+
+    @staticmethod
+    def _randomize_value(value, interval):
+        return random.randint(
+            int(value * (1 - interval)), int(value * (1 + interval))
+        )
+
+    def _set_shuffle_intervals(self):
+        if self.max_scene_repetition_episodes > 0:
+            self._max_rep_episode = self._randomize_value(
+                self.max_scene_repetition_episodes,
+                self.repetition_rand_interval,
+            )
+        else:
+            self._max_rep_episode = None
+
+        if self.max_scene_repetition_steps > 0:
+            self._max_rep_step = self._randomize_value(
+                self.max_scene_repetition_steps, self.repetition_rand_interval
+            )
+        else:
+            self._max_rep_step = None
+
+    def _switch_scene_if(self):
```

Comment:
Question here: would it be switching to frequently at the later stage of training? with potentially less than 100 episode per scene switch? Will it help to incorporate both count schemes, like a logical AND?

Reply:
Shuffling on steps is consistent from an optimization standpoint, you swap scenes every N parameters updates, which is why I like it :)

I don't think you can shuffle scenes "too often".  Ideally, we just randomly sample a new episode irrespective of the scene it is in, but this incurs the non-trivial cost of swapping the scene way too often.